app/controllers/mixins/breadcrumbs_mixin.rb

```diff
@@ -33,33 +34,63 @@ def data_for_breadcrumbs
 
-          # Ancestry parents breadcrumbs
+        # On special page (tagitems, politems, etc.)
+        if @tagitems || @politems || @ownershipitems || @retireitems
+          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems, true)) unless options[:hide_special_item]
+          breadcrumbs.push(:title => @right_cell_text)
+        else
+          # Ancestry parents breadcrumbs (only in services)
           breadcrumbs.concat(ancestry_parents(options[:ancestry], options[:record_info], options[:record_title])) if options[:ancestry]
 
-          breadcrumbs.push(:title => options[:record_info][options[:record_title]], :key => x_node_right_cell)
-          breadcrumbs.push(:title => title_for_breadcrumbs) if options[:show_header] && @right_cell_text
-        else
-          # Append breadcrumb from the title of right cell
-          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems)) unless options[:hide_special_item]
-          breadcrumbs.push(:title => title_for_breadcrumbs) if action_breadcrumb?
+          # Try to prepare TreeNode on every item outside the root
+          if x_node != 'root'
+            if options[:include_record] && options[:record_info].present?
+              # Include items (for trees, which does not have final nodes: VMs etc.)
+              node = TreeNode.new(options[:record_info])
+            else
+              # Try to create treenode from the x_node
+              model, model_id, _ = TreeBuilder.extract_node_model_and_id(x_node) if x_node
+              record = model.try(:constantize).try(:find, model_id)
+
+              node = TreeNode.new(record) if record
+            end
+          end
+
+          # Overrides: if x_root then show title_for_breadcrumbs
+          #
+          # Priority 1. Include record
+          #          2. TreeNode.text
+          #          3. Item from tree (if it's built)
+          #          4. Fallbacks (sent text, title of the record, title_for_breadcrumbs)
+
+          key = node ? node.key : x_node
+          title = if x_node == 'root'   # After switching accordion, there is no way how to get root text
+                    title_for_breadcrumbs
+                  elsif node            # Use node's text and key
+                    node.text
+                  elsif @trees.present? # Select a node when coming from different page (the tree is built)
+                    build_breadcrumb_from_tree.try(:[], :title)
+                  else                  # Last clicked tree node label
+                    @x_node_text.try(:[], x_active_tree)
+                  end
+
+          breadcrumbs.push(:title => title, :key => key) if title
+
+          # Action header breadcrumb (i.e. edit, etc.)
+          last_title = title_for_breadcrumbs || @title
+          breadcrumbs.push(:title => last_title) if action_breadcrumb? && last_title && title != last_title
```

Comment:
Is it possible to somehow incorporate this `last_title` into the `title` and so not to have to call `breadcrumbs.push` twice?

Reply:
They should be called twice, they are two separate breadcrumbs.